picard/ui/metadatabox.py

```diff
@@ -585,14 +585,19 @@ def _update_tags(self, new_selection=True, drop_album_caches=False):
         top_tags = config.setting['metadatabox_top_tags']
         top_tags_set = set(top_tags)
 
+        format_settings = {
```

Comment:
As per previous comment, since it is f_s_m that knows what settings it needs, and to prevent code duplication, I think this should be `format_settings = {}` or even `format_settings_cache` and this (mutable dict) object should then be populated inside `format_specific_metadata`.

Reply:
I don't like this passing the settings back, that makes IMHO for a pretty confusing interface. I'd usually not expect a function to manipulate the passed settings. It also makes validating the settings more complicated. In the current case the function can just check if settings is not None, then initialize it. In the other case it needs to check for each key. The goal is exactly to keep this settings handling basically out of the function.

For the same reason I would not introduce another function to handle that. We try to reduce the call complexity here.

Reply:
> I don't like this passing the settings back, that makes IMHO for a pretty confusing interface.

I don't think this is any more confusing than requiring someone to know what settings are needed and duplicating the code.

>I'd usually not expect a function to manipulate the passed settings.

I agree that this is less intuitive - but same comment as above.

> It also makes validating the settings more complicated. 

It depends. The code I have suggested is no more complicated than current - but it doesn't validate that the settings include what is needed - because it assumes that an empty dict is passed in. But if you want to validate it would be a one line change:
```
 def format_specific_metadata(self, metadata, tag, settings={}):
        if ("write_id3v23" not in settings or "id3v23_join_with" not in settings):
            config = get_config()
            settings["write_id3v23"] = config.setting["write_id3v23"]
            settings["id3v23_join_with"] = config.setting["id3v23_join_with"]
```
> In the current case the function can just check if settings is not None, then initialize it. In the other case it needs to check for each key. The goal is exactly to keep this settings handling basically out of the function.

In that case, settings should NOT be given a default of None.
 
> For the same reason I would not introduce another function to handle that. We try to reduce the call complexity here.

I agree we should **not** introduce another function call if it gets called once per tag per file. But if it is only called once per metadatabox update, what is the problem of having a helper function? If the new thread-safe config functions have a performance impact, then IMO there needs to be a mitigation of that provided as a helper.

Reply:
Ok, I gave this a try with passing back the config values back to a cache object. Probably the biggest advantage is indeed avoiding duplication. Also we could have different implementations of `File.format_specific_metadata` that will need different config values.

This PR needs some tests added for both the new `cached_settings()`  and `File.format_specific_metadata`, otherwise this seems to be fine to me.

>  If the new thread-safe config functions

It's not only the thread safe config functions, it is config access in general. `config.settings['foo']`  is comparable slow, that's nothing new. If you have config.setting access in the ID3 format_specific_metadata then the above use case for ID3v2.3 goes up from ~0.6 seconds to ~2.0 seconds on my system. If you additionally have `get_config`  call in it it goes up to ~2.2 seconds. That's big, but the biggest bottleneck is general config access.

Reply:
> Ok, I gave this a try with passing back the config values back to a cache object. Probably the biggest advantage is indeed avoiding duplication. Also we could have different implementations of `File.format_specific_metadata` that will need different config values.

Good point - and we need to handle cases where the metadata being summarised includes a variety of different formats. And I think the current implementation does that.

> This PR needs some tests added for both the new `cached_settings()` and `File.format_specific_metadata`, otherwise this seems to be fine to me.

Thanks to @phw Philipp for taking this PR on from my initial code - I really **must** try to find the time to set up a development environment so that Philipp doesn't have to cope with my comments from the sidelines and generating additional tests that I haven't provided.

> > If the new thread-safe config functions
> 
> It's not only the thread safe config functions, it is config access in general. `config.settings['foo']` is comparable slow, that's nothing new. If you have config.setting access in the ID3 format_specific_metadata then the above use case for ID3v2.3 goes up from ~0.6 seconds to ~2.0 seconds on my system. If you additionally have `get_config` call in it it goes up to ~2.2 seconds. That's big, but the biggest bottleneck is general config access.

Hmmmm...

Perhaps what we need is a more general config cache that is applied automatically, with cache entries saved with a time-stamp and ignored/updated if the time-stamp is older than (say) 5 seconds. I cannot remember what exactly the Config issues was that required thread specific Config objects, but if we had a global cache, would that allow us to move back to a single global Config object and just deal with any issue of locking to prevent simultaneous updates between threads and / or multiple Picard instances?